### PR TITLE
[spirv] Expose -Oconfig for running custom optimizer recipes

### PR DESCRIPTION
**Pull request by ehsannas**
Also updated SPIR-V Tools to get the latest API.

### CODE REVIEW

**Comment by antiagainst** [comment]
Oh, we should also update the dochttps://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/SPIR-V.rst#optimization

File: include/dxc/Support/HLSLOptions.td [hunk 1]
```diff
@@ -262,6 +262,8 @@ def fspv_target_env_EQ : Joined<["-"], "fspv-target-env=">, Group<spirv_Group>,
   HelpText<"Specify the target environment: vulkan1.0 (default) or vulkan1.1">;
 def Wno_vk_ignored_features : Joined<["-"], "Wno-vk-ignored-features">, Group<spirv_Group>, Flags<[CoreOption, DriverOption, HelpHidden]>,
   HelpText<"Do not emit warnings for ingored features resulting from no Vulkan support">;
+def Oconfig : CommaJoined<["-"], "Oconfig=">, Group<spirv_Group>, Flags<[CoreOption]>,
+  HelpText<"Custom Optimization Configuration: Comma-separated list of configurations">;
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply include/dxc/Support/HLSLOptions.td]: "... (see http://khr.io/hlsl2spirv#optimization)"

[the remark this answers is not in the document]
Comment (ehsannas) [reply include/dxc/Support/HLSLOptions.td]: Done

File: lib/DxcSupport/HLSLOptions.cpp [hunk 2]
```diff
@@ -568,6 +568,26 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
   }
 
   opts.SpvTargetEnv = Args.getLastArgValue(OPT_fspv_target_env_EQ, "vulkan1.0");
+
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply lib/DxcSupport/HLSLOptions.cpp]: Done

File: lib/DxcSupport/HLSLOptions.cpp [hunk 3]
```diff
@@ -568,6 +568,26 @@ int ReadDxcOpts(const OptTable *optionTable, unsigned flagsToInclude,
...
+  // Handle -Oconfig=<comma-separated-list> option.
+  uint32_t numOconfigs = 0;
+  for (const Arg *A : Args.filtered(OPT_Oconfig)) {
+    ++numOconfigs;
+    if (numOconfigs > 1) {
+      errors << "-Oconfig should not be specified more than once";
+      return 1;
+    }
+    for (const auto opt : {OPT_O0, OPT_O1, OPT_O2, OPT_O3, OPT_O4}) {
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply lib/DxcSupport/HLSLOptions.cpp]: Done

File: tools/clang/include/clang/SPIRV/EmitSPIRVOptions.h [hunk 4]
```diff
@@ -50,6 +50,7 @@ struct EmitSPIRVOptions {
   spirv::LayoutRule cBufferLayoutRule;
   spirv::LayoutRule tBufferLayoutRule;
   spirv::LayoutRule sBufferLayoutRule;
+  llvm::SmallVector<llvm::StringRef, 4> oConfig;
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/include/clang/SPIRV/EmitSPIRVOptions.h]: Done

File: docs/SPIR-V.rst [hunk 5]
```diff
@@ -347,7 +347,65 @@ Optimization
...
+* ``--scalar-replacement[=<n>]``
+* ``--set-spec-const-default-value "<spec id>:<default value> ..."``
+* ``--simplify-instructions``
+* ``--strength-reduction``
+* ``--strip-debug``
+* ``--strip-reflect``
+* ``--vector-dce``
+* ``--workaround-1209``
+* ``--unify-const``
+
```
Comment (antiagainst) [inline docs/SPIR-V.rst]: Besides, there are two special batch options; each stands for a recommended recipe by itself:

* `-O`: A bunch of passes in an appropriate order that attempt to improve performance of generated code. Same as `spirv-opt -O`. Also same as SPIR-V CodeGen's default recipe.
* `-Os`: A bunch of passes in an appropriate order that attempt to reduce the size of the generated code. Same as `spirv-opt -Os`.

So if you want to run loop unrolling additionally after the default optimization recipe, you can specify `-Oconfig=-O,--loop-unroll`.

For the whole list of accepted passes and details about each one, please see `spirv-opt`'s help manual (`spirv-opt --help), or the SPIRV-Tools [optimizer header file](https://github.com/KhronosGroup/SPIRV-Tools/blob/master/include/spirv-tools/optimizer.hpp).

File: docs/SPIR-V.rst [hunk 6]
```diff
@@ -347,7 +347,65 @@ Optimization
...
 between optimization levels greater than zero; they will all invoke the same
-optimization recipe. This may change in the future.
+optimization recipe. If you want to run a custom optimization recipe, you can do
+so using ``-Oconfig=`` and specifying a comma-separated list of your desired passes.
+
+For example, you can specify: ``-Oconfig=--loop-unroll,--eliminate-dead-variables,--scalar-replacement=300``.
+
+Here is a list of valid passes:
+
+* ``--ccp``
```
Comment (antiagainst) [inline docs/SPIR-V.rst]: Sorry, but we should clean this list a little bit. Some of the passes are superseded by others; we should not recommend them. Also we should avoid mentioning `--legalize-hlsl`, which is not supposed to be invoked dedicatedly by the developer, and put `-O` and `-Os` separately.

**Pull request merged**